File: src/main/java/com/pretty/platform/catalog/interfaces/rest/resources/ProductResource.java
package com.pretty.platform.catalog.interfaces.rest.resources;

import io.swagger.v3.oas.annotations.media.Schema;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

/**
 * Resource representing a product
 */
@Schema(description = "Product response")
public record ProductResource(
    @Schema(description = "Product ID", example = "550e8400-e29b-41d4-a716-446655440000")
    UUID id,

    @Schema(description = "Product title", example = "Wireless Headphones")
    String title,

    @Schema(description = "Product description", example = "High-quality wireless headphones")
    String description,

    @Schema(description = "Product brand", example = "Sony")
    String brand,

    @Schema(description = "Product categories", example = "[\"Electronics\", \"Audio\"]")
    List<String> categories,

    @Schema(description = "Product tags", example = "[\"wireless\", \"bluetooth\"]")
    List<String> tags,

    @Schema(description = "Product image URLs", example = "[\"https://example.com/image1.jpg\"]")
    List<String> imageUrls,

    @Schema(description = "Creation timestamp")
    LocalDateTime createdAt,

    @Schema(description = "Last update timestamp")
    LocalDateTime updatedAt
) {}

File: internal/products/interfaces/rest/resources/product_resource.go
package resources

import "time"

type ProductResource struct {
	ID          string    `json:"id" example:"550e8400-e29b-41d4-a716-446655440000"`
	ImageURL    string    `json:"image_url" example:"https://example.com/image.jpg" validate:"required,url"`
	Title       string    `json:"title" example:"Product Title" validate:"required,min=1,max=255"`
	Price       float64   `json:"price" example:"99.99" validate:"required,min=0"`
	Description string    `json:"description" example:"Product description" validate:"max=1000"`
	Category    string    `json:"category" example:"Electronics" validate:"required"`
	Quantity    int       `json:"quantity" example:"10" validate:"required,min=0"`
	CreatedAt   time.Time `json:"created_at" example:"2023-01-01T00:00:00Z"`
	UpdatedAt   time.Time `json:"updated_at" example:"2023-01-01T00:00:00Z"`
}

type CreateProductResource struct {
	ImageURL    string  `json:"image_url" example:"https://example.com/image.jpg" validate:"required,url"`
	Title       string  `json:"title" example:"Product Title" validate:"required,min=1,max=255"`
	Price       float64 `json:"price" example:"99.99" validate:"required,min=0"`
	Description string  `json:"description" example:"Product description" validate:"max=1000"`
	Category    string  `json:"category" example:"Electronics" validate:"required"`
	Quantity    int     `json:"quantity" example:"10" validate:"required,min=0"`
}

type UpdateProductResource struct {
	ImageURL    *string  `json:"image_url,omitempty" example:"https://example.com/image.jpg" validate:"omitempty,url"`
	Title       *string  `json:"title,omitempty" example:"Product Title" validate:"omitempty,min=1,max=255"`
	Price       *float64 `json:"price,omitempty" example:"99.99" validate:"omitempty,min=0"`
	Description *string  `json:"description,omitempty" example:"Product description" validate:"omitempty,max=1000"`
	Category    *string  `json:"category,omitempty" example:"Electronics" validate:"omitempty"`
	Quantity    *int     `json:"quantity,omitempty" example:"10" validate:"omitempty,min=0"`
}

type ErrorResponse struct {
	Error string `json:"error" example:"error message"`
}
